txpipe/utils/splitters.py

```diff
@@ -0,0 +1,207 @@
+        for col in self.columns:
+            group[col][s:e] = data[col]
+
+        # Update overall index
+        self.index[b] = e
+
+    def _size_check(self, b, e):
+        n = self.bin_sizes[b]
+        if e > n:
+            raise ValueError(f"Too much data added bin {b}: got {e}, expected max {n}")
+
+    def finish(self, bins=None):
+        """
+        Finish up by checking that the right amount of data has been written to the file.
+
+        If running in parallel, only bins used by this process will have the correct sizes
+        in this object.  In that case, specify my_bins for the list of bins this process
+        should check
+
+        Parameters
+        ----------
+        bins: list or None
+            The list of bins that this process has saved data for.
+        """
+        if bins is None:
+            bins = self.bins
+
+        for b in bins:
+            c = self.bin_sizes[b]
+            n = self.index[b]
+            if c != n:
+                raise ValueError(
+                    f"Count error in bin {b}: expected {c} but copied in {n}"
+                )
+
+
+class DynamicSplitter(Splitter):
+    """
+    This version of the splitter dynamically adjusts the sizes of the columns using
+    chunking, so you don't need to know the sizes in advance.  The cost is that it
+    can't be used in parallel.
+
+    The sizes pased to the initialization in this case represent an initial guess of
+    the column sizes; data will be resized above this.
+
+    See the Splitter docstring for more detail.
+    """
+
+    def __init__(self, group, name, columns, bin_sizes, dtypes=None):
+        """Create a dynamic splitter.
+
+
+        Parameters
+        ----------
+        group: h5py.Group
+            The group where the output data will be written
+        name: str
+            The base name of the different bins to write
+        columns: list
+            The str names of the columns to be split
+        bin_sizes: dict
+            Maps bin_name (can be anything printable) to an initial guess
+            of the size.  Bins will be expanded as needed above this.
+        dtypes: dict or None
+            Maps bins to HDF5 data types for the output columns.  Bins default
+            to 8 byte floats if not found in this.
+        """
+        super().__init__(group, name, columns, bin_sizes, dtypes=dtypes)
+
+    def _setup_columns(self, dtypes):
+        # same as in the parent class except we make them extensible by
+        # setting maxshape
+        for b, sz in self.bin_sizes.items():
+            sub = self.subgroups[b]
+            for col in self.columns:
+                dt = dtypes.get(col, "f8")
+                sub.create_dataset(col, (sz,), dtype=dt, maxshape=(None,))
+
+    def _size_check(self, b, e):
+        n = self.bin_sizes[b]
+
+        # Expand the columns by 50% if needed
+        while e > n:
+            sub = self.subgroups[b]
+            new_size = int(n * 1.5)
```

Comment:
I don't know how often e > n*1.5, but if that's ever a use case, you're probably better off either doing `new_size = int(e * 1.5)` to give space above the current new size requirement, or at least `new_size = max(e, int(n*1.5))` to avoid multiple resizes.

With either of these choices, you can switch the while to an if.

Reply:
Ha, you're absolutely right, that makes much more sense. Weird what blind spots I get sometimes.  I'm changing to return early if the size is already okay.